SimPEG/electromagnetics/static/utils/static_utils.py

```diff
@@ -687,29 +695,345 @@ def plot_3d_pseudosection(
+                )
+
+            marker['color'] = plot_vec[k]
+            scatter_data = [
+                grapho.Scatter3d(
+                    x=locations[k, 0],
+                    y=locations[k, 1],
+                    z=locations[k, 2],
+                    customdata=dvec[k],
+                    mode='markers',
+                    marker=marker,
+                )
+            ]
+
+        fig = grapho.Figure(data=scatter_data)
+
+        fig.update_layout(
+            scene=dict(
+                xaxis=dict(title='X[m]'),
+                yaxis=dict(title='Y[m]'),
+                zaxis=dict(title='Z[m]')
+            ),
+            scene_camera=dict(eye=dict(x=1.5, y=-1.5, z=1.5))
+        )
+        if layout_opts is not None:
+            fig.update_layout(**layout_opts)
+
+        return fig
+
+
+
+
+
+#########################################################################
+#                      GENERATE SURVEYS
+#########################################################################
+
+
+def generate_survey_from_abmn_locations(
+    *,
+    locations_a=None,
+    locations_b=None,
+    locations_m=None,
+    locations_n=None,
+    data_type=None,
+    output_sorting=False
+):
+    """
+    Use A, B, M and N electrode locations to construct a 2D or 3D DC/IP survey.
+
+    Parameters
+    ----------
+    
+    locations_a : numpy.array
+        An (n, dim) numpy array containing A electrode locations
+    locations_b : None or numpy.array
+        An (n, dim) numpy array containing B electrode locations. If None,
+        we assume all sources are Pole sources.
+    locations_m : numpy.array
+        An (n, dim) numpy array containing M electrode locations
+    locations_n : numpy.array
+        An (n, dim) numpy array containing N electrode locations. If None,
+        we assume all receivers are Pole receivers.
+    data_type : str
+        Must be one of {'volt', 'apparent_conductivity', 'apparent_resistivity', 'apparent_chargeability'}
+    output_sorting : bool
+        This option is used if the ABMN locations are sorted during the creation of the survey
+        and you would like to sort any data vectors associated with the electrode locations.
+        If False, the function will output a SimPEG.electromagnetic.static.survey.Survey object.
+        If True, the function will output a tuple containing the survey object and a numpy array
+        (n,) that will sort the data vector to match the order of the electrodes in the survey.
+
+
+    Returns
+    -------
+    survey
+        A SimPEG.electromagnetic.static.survey.Survey object
+    sort_index
+        A numpy array which defines any sorting that took place when creating the survey
+
+
+    """
+    
+    if locations_a is None:
+        raise AssertionError("Locations for A electrodes must be provided.")
+    if locations_m is None:
+        raise AssertionError("Locations for M electrodes must be provided.")
+
+    assert data_type.lower() in [
```

Comment:
do not use `assert`, it is in general bad practice when checking the type as it's possible to turn asserts off. Instead throw a `ValueError`

Reply:
done